Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h

```diff
@@ -74,32 +81,131 @@ class EMCALCalibExtractor
   template <typename... axes>
   o2::emcal::BadChannelMap calibrateBadChannels(boost::histogram::histogram<axes...>& hist)
   {
-    // calculate the mean and sigma of the mEsumHisto
-    auto fitValues = o2::utils::fitBoostHistoWithGaus<double>(hist);
-    double mean = fitValues.at(1);
-    double sigma = fitValues.at(2);
-
-    // calculate the "good cell window from the mean"
-    double maxVal = mean + mSigma * sigma;
-    double minVal = mean - mSigma * sigma;
+    double time1 = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now().time_since_epoch()).count();
+    std::map<int, std::pair<double, double>> slices = {{0, {0.1, 0.3}}, {1, {0.3, 0.5}}, {2, {0.5, 1.0}}, {3, {1.0, 4.0}}};
+
+    // get all ofthe calibration information that we need in a struct
+    BadChannelCalibInfo calibrationInformation = buildHitAndEnergyMean(slices, hist);
+
     o2::emcal::BadChannelMap mOutputBCM;
     // now loop through the cells and determine the mask for a given cell
-    for (int cellID = 0; cellID < 17664; cellID++) {
-      double E = hist.at(cellID);
-      // if in the good window, mark the cell as good
-      // for now we won't do warm cells - the definition of this is unclear.
-      if (E == 0) {
+
+#if (defined(WITH_OPENMP) && !defined(__CLING__))
+    if (mNThreads < 1) {
+      mNThreads = std::min(omp_get_max_threads(), mNcells);
+    }
+    LOG(info) << "Number of threads that will be used = " << mNThreads;
+#pragma omp parallel for num_threads(mNThreads)
+#else
+    LOG(info) << "OPEN MP will not be used for the bad channel calibration";
+    mNThreads = 1;
+#endif
+
+    for (int cellID = 0; cellID < mNcells; cellID++) {
+      if (calibrationInformation.energyPerHitMap[0][cellID] == 0) {
+        LOG(debug) << "Cell " << cellID << " is dead.";
         mOutputBCM.addBadChannel(cellID, o2::emcal::BadChannelMap::MaskType_t::DEAD_CELL);
-      } else if (E < maxVal || E > minVal) {
-        mOutputBCM.addBadChannel(cellID, o2::emcal::BadChannelMap::MaskType_t::GOOD_CELL);
       } else {
-        mOutputBCM.addBadChannel(cellID, o2::emcal::BadChannelMap::MaskType_t::BAD_CELL);
+        bool failed = false;
+        for (auto& [index, slice] : slices) {
+          auto ranges = calibrationInformation.goodCellWindowMap[index];
+          auto meanPerCell = calibrationInformation.energyPerHitMap[index][cellID];
```

Comment:
Be careful with variable names (`index` in this case) - not that you redefine existing names